### ISSUE TITLE
make jimp non-blocking

### ISSUE THREAD

**Issue by bchr02** [oliver-moran/jimp#229]
Any change jimp could be updated so it's non-blocking? Using jimp I created a program to change the quality on a large set of images and the whole time the process was running it blocked the event loop preventing other things from running. Also, the memory consumption on the node process skyrocketed to over 500MB. A quick review of your code shows me that everything is processed in memory by use of Buffers (and I don't mean Buffered Streams either). Image files are large and would be better processed in [Streams](https://nodejs.org/api/stream.html). This would prevent the event loop from getting blocked and keep memory consumption down.

**Comment by Iwasawafag** [comment]
I would say it's unlikely going to happen in the near future. We could change methods of Jimp so it would iterate on pixels in chunks rather all at once, maybe by measuring time, but that would only solve half of the problem. And if multiple "asynchronous" operations will be started in the same time event loop will struggle anyway. Image encoding/decoding on the other half is handled by third party modules and iirc jpeg-js and bmp-js are synchronous-only. (Also jpeg-js and pngjs modules are the ones to blame for high memory usage #153). 

Stream support mostly limited by the same factors - encoding/decoding algorithms. But it will also likely break methods like color normalization that rely on the histogram of the entire image. Maybe ImageMagick or libvips is more suitable to your needs

**Comment by thisconnect** [comment]
What if "all" jimp methods use Promises? Assuming not the entire api is promise based. Would that give the event loop some room to breath?

**Comment by Iwasawafag** [comment]
No, every CPU expensive operation would still block the thread. Maybe something will be able to run in between two Jimp operations, because promises are always resolved asynchonously, but that's about it

**Comment by aurium** [comment]
@Iwasawafag is right! And what about to use the [node-webworker-threads module](https://github.com/audreyt/node-webworker-threads) to use multi-core CPUs? I don't think it gonna be simple, but looks possible, (not exactly) as it is used by [synaptic](https://github.com/cazala/synaptic/).
```js
if (os.cpus().length > 2) Threads.create().eval(()=>{ ... }, userCallback);
```

**Comment by bchr02** [comment]
I solved the blocking by creating child processes to run the jimp code in.

For example:
```javascript
// ./main_app.js
const path = require('path');
const spawn = require('child_process').spawn;

const compressImages = (folder, quality, callback) => {
	let err = '';

	const j = spawn('node.exe', 
		['--max-old-space-size=1000', 'child_jimp.js', quality, folder], 
		{cwd: path.join(__dirname, '../syncProcesses')
	});

	j.stdout.on('data', (data) => {
		console.log(data.toString('utf8').replace(/[\n\r]/g, ''));
	});

	j.stderr.on('data', (data) => {
		err += data.toString('utf8');
	});

	j.on('close', (code) => {
		console.log('successfully optimized all images');
		callback(err);
	});
};

// Here I run an Express server which provides an endpoint to submit files. 
// Those files are saved into a temp folder. Then I do something like this:
compressImages(tempfolder, 60, ()=>{
	delete_tmp_folder();
});
```

```javascript
// ./syncProcesses/child_jimp.js

if(process.argv.length < 4) throw("Not enough arguments provided to child_jimp.js");

const fs = require('fs');
const jimp = require("jimp");
const quality = parseInt(process.argv[2], 10);
const folder = process.argv[3];

fs.readdir(folder, (err, files) => {
    if(err) throw(err);
    files.forEach(fn => {
        const file = folder + '/' + fn;
        const optimized_file = folder + '/optimized_' + fn;
        jimp.read(file, function (err, file) {
            if(err) throw(err);
            file.quality(quality) // set JPEG quality
                .write(optimized_file, (err) => { // save
                    if(err) throw(err);
                    console.log('successfully optimized image ' + optimized_file);
                });
        });
    });
});
```

**Comment by bchr02** [comment]
> I solved the blocking by creating child processes to run the jimp code in.

It would be nice if jimp could provide the means of doing this internally.

**Comment by bchr02** [comment]
@oliver-moran @Iwasawafag @aurium What are your thoughts on implementing [something like this](https://github.com/oliver-moran/jimp/issues/229#issuecomment-274972654) within jimp?

**Comment by Iwasawafag** [comment]
I don't know about this. 

I would love to be able to pass raw buffers with encoded images or raw RGBA pixel data to avoid writing temp files to disk. Tossing data back and forth over stdin/stdout for every operation would be a waste of time, so better approach is to change syntax to declarative style, where you specify order of operations from the start and then let remote process do the job similarly to API `gm` module provides. But that comes with its own limitations, for example, you can't put a condition in the middle of the chain depending on intermediate state of the image. 

I don't like idea of breaking existing API and moving to declarative syntax. If we had to go with multiprocess approach I'd rather wrote an adapter to gm module, but there are plenty of differences in feature set and supported formats between GraphicsMagick/ImageMagick and Jimp so that can't be done over night and I don't think gm was designed to support anything other than GM/IM.

Multithreading sounds like a better solution, it's supposed to be able to work on the same memory but from what I gathered in node could only be done with a module, [most popular of which](https://www.npmjs.com/package/webworker-threads)(?) requires some binaries to be compiled and that goes against the idea of having "pure javascript" image processor.

And lets not forget about browser. Browser version has access to workers, but can't spawn child processes. Also it's designed around the idea of image processing inside the worker and I'm not sure if one worker can spawn more workers. 

Maybe this should be done as a separate module intended for use in Node.js only, but then again, if compiling native modules is not a problem – there are faster alternatives should be available and I don't see appeal of pure javascript solution anymore

**Comment by bchr02** [comment]
I'm not suggesting we implement it by reading from disk. The example was somewhat unique to my situation. Also, I probably should change the example so that child_jimp.js only does the jimp functions and returns a success or an error and then do the folder loop within the main_app.js. Regardless, what I wanted you to take away from the example was how relatively easily we can implement the offloading of blocking tasks to child processes.

> I don't like idea of breaking existing API and moving to declarative syntax.

I agree and don't suggest we change the API I suggest we add to it. For example, in the situation where someone want to simply change the quality of a file they could make one method call, something like `jimp.qualityasync(file, [file_to], callback);` which can abstract the offloading of the blocking task to a sub process.

**Comment by aurium** [comment]
`webworker-threads` requires some binaries to be compiled, however it is necessary only at Node.js context. At Browser's we can use WebWorkers. Both solutions are multi-core friendly and we may consider to do all the manipulation job in a worker using messages to send commands and returning values.

The load and save will be increased by the thread communication, however i don't know how much it will affect the r/w time and if it will be better or not to all the job. I would bet it can be good.

**Comment by thisconnect** [comment]
I am concerned that jimp already is quite big with many dependencies (). batch processing jimp could be something that is built on top of jimp and optimized for the environment. I have never used jimp in the browser, but I understand that it makes sense for some people. 

jimp itself should not blow up IMO. We could add jimp-webworker, jimp-spawn, jimp-cli. Unfortunatelly it seems https://github.com/jimp (for a jimp github organization) is already taken

**Comment by thisconnect** [comment]
I don't get what webworker-threads is https://github.com/audreyt/node-webworker-threads/issues/146

**Comment by aurium** [comment]
>  jimp itself should not blow up IMO. We could add jimp-webworker, jimp-spawn, jimp-cli. Unfortunatelly it seems https://github.com/jimp (for a jimp github organization) is already taken

I second you!

> I don't get what webworker-threads is audreyt/node-webworker-threads#146

Sorry, show you the `Threads` usage, however the focus should be on `Worker` usage to have the same API on Node.js and browsers.

**Comment by Iwasawafag** [comment]
> `webworker-threads` requires some binaries to be compiled, however it is necessary only at Node.js context. At Browser's we can use WebWorkers. Both solutions are multi-core friendly and we may consider to do all the manipulation job in a worker using messages to send commands and returning values.

I just happened to come across `Readme.md` again and noticed that it says the following:

> An image processing library for Node written entirely in JavaScript, with zero external or native dependencies.

Having no native dependencies is selling point, as sometimes those fail to compile and on Windows it's common to not even have a compiler installed.

So yeah, I prefer variant with second module that would implement the same interface, but would execute every operation in a separate thread, what @thisconnect suggested.

As for creating organization and multiple repositories – maybe monorepo is the way to go? I tried Lerna and first impression was okay-ish. We already have about 300MB of test images committed here which isn't going anywhere and I don't like idea of having more of those in other repositories. At the same time having one test case for all plugins/wrappers sounds very appealing to me, especially considering that most of operations would be the same and should produce the same output.
